trin-execution/src/trie_walker/mod.rs

```diff
@@ -98,23 +88,26 @@ impl<TrieDB: DB> TrieWalker<TrieDB> {
                     return Err(anyhow::anyhow!("Node not found in the database"));
                 }
             };
-            let decoded_node = decode_node(&mut value.as_ref())?;
-            match decoded_node {
-                Node::Leaf(_) | Node::Extension(_) | Node::Branch(_) => {
-                    self.stack.push(TrieProof {
-                        path,
-                        proof: [partial_proof, vec![value.into()]].concat(),
-                    });
-                }
-                // can't be a hash node because we just decoded it
-                Node::Empty | Node::Hash(_) => (),
+
+            // check that node decodes correctly and to correct variant
+            if matches!(
+                decode_node(&mut value.as_ref())?,
+                Node::Empty | Node::Hash(_)
+            ) {
+                return Ok(());
             }
+
+            self.stack.push(TrieProof {
+                path,
+                proof: [partial_proof, vec![value]].concat(),
```

Comment:
I think there is no reason to use `concat()`, as I believe that it would iterate and clone partial_proof (it should be cheap, but still not needed). We can:

```rust
let mut proof = partial_proof
proof.push(value);
self.stack.push(TrieProof { path, proof });
```

or if you prefer:
```rust
partial_proof.push(value);
self.stack.push(TrieProof {
    path,
    proof: partial_proof,
});
```